lte/gateway/c/session_manager/SessionStateEnforcer.cpp

```diff
@@ -136,15 +138,241 @@ void SessionStateEnforcer::handle_session_init_rule_updates(
+      map_element->second.erase(session_element--);
+      MLOG(MINFO) << "Successfully terminated session " << session_id;
+      /* If it is last session terminated and no session left for this IMSI
+       * remove the imsi as well
+       */
+      if (map_element->second.size() == 0) {
+        session_map.erase(imsi);
+        MLOG(MINFO) << "All sessions terminated for IMSI " << imsi;
+      }
+      return;
+    }
+  }  // end of for loop
+  // reached end of map and no session found.
+  MLOG(MINFO) << "No session found to complete termination"
+              << " session id " << session_id;
+  return;
+}
+
+/*Removing all associated rules to session*/
+void SessionStateEnforcer::m5g_remove_rules_from_session_for_termination(
+    const std::string& imsi, const std::unique_ptr<SessionState>& session,
+    const std::string& dnn, SessionStateUpdateCriteria& uc) {
+  // remove all rules;
+  MLOG(MINFO) << "Will be removing all associated rules of IMSI " << imsi;
+  session->remove_all_rules();
+  auto ip_addr = session->get_config()
+                     .rat_specific_context.m5gsm_session_context()
+                     .pdu_address()
+                     .redirect_server_address();
+  // Update to UPF
+  SessionState::SessionInfo sess_info;
+  sess_info.imsi    = imsi;
+  sess_info.ip_addr = ip_addr;
+  session->sess_infocopy(&sess_info);
+  pipelined_client_->set_upf_session(sess_info, call_back_upf);
+  return;
+}
+
+/* This function will change the state of respective PDU session,
+ * upon receving message or notification from UPF
+ * Defined to handle ACTIVE or INACTIVE state of session
+ * based on UPF message to SMF
+ * imsi - from UPF handler to find respective SessionMap
+ * seid - session context id to find respective session
+ * new_state - state change required w.r.t. UPF message
+ */
+void SessionStateEnforcer::m5g_update_session_state_message_from_upf(
+    const std::string& imsi, uint32_t teid, uint32_t version,
+    SessionFsmState new_state) {
+  auto session_map = session_store_.read_sessions({imsi});
+  /* Search with session search criteria of IMSI and session_id and
+   * find  respective sesion to operate
+   */
+  SessionSearchCriteria criteria(imsi, IMSI_AND_TEID, teid);
+  auto session_it = session_store_.find_session(session_map, criteria);
+  if (!session_it) {
+    MLOG(MERROR) << "No session found in SessionMap for IMSI " << imsi
+                 << " with teid " << teid;
+    return;
+  }
+  MLOG(MWARNING) << "te id (" << teid << ") found ..updating to AMF";
+  auto& session       = **session_it;
+  auto session_update = SessionStore::get_default_session_update(session_map);
+  SessionStateUpdateCriteria& session_uc =
+      session_update[imsi][session->get_session_id()];
+  session->set_fsm_state(new_state, session_uc);
+  uint32_t cur_version = session->get_current_version();
+  session->set_current_version(++cur_version, session_uc);
+  MLOG(MINFO) << "state of session changed to "

```

Comment:
Again I think this information is logged in `set_fsm_state`. If we also want to log version changes, we should put that log in `set_current_version` since we seem to always log it afterwards.

Reply:
removed